aggregator/src/aggregator/aggregation_job_writer.rs

```diff
@@ -593,6 +570,442 @@ impl<const SEED_SIZE: usize, Q: CollectableQueryType, A: vdaf::Aggregator<SEED_S
         )
         .await?;
 
-        Ok(unwritable_report_ids)
+        Ok(indexed.unwritable_report_ids)
+    }
+}
+
+/// Generic callback used in the internals of aggregation job writers. Different implementations are
+/// used when creating new aggregation jobs versus updating existing aggregation jobs.
+trait BatchUpdateCallback {
+    /// This takes one batch and an iterator of aggregation jobs, and returns that batch updated to
+    /// reflect changes due to the aggregation job. Particularly, it will update the
+    /// `outstanding_aggregation_jobs` counter and the `client_timestamp_interval`.
+    fn update_batch<'a, const SEED_SIZE: usize, Q, A, RA>(
+        &self,
+        batch: Batch<SEED_SIZE, Q, A>,
+        aggregation_jobs: impl Iterator<
+            Item = (
+                &'a AggregationJob<SEED_SIZE, Q, A>,
+                &'a [Cow<'a, RA>],
+                &'a [usize],
+            ),
+        >,
+    ) -> Result<Batch<SEED_SIZE, Q, A>, Error>
+    where
+        Q: CollectableQueryType,
+        A: vdaf::Aggregator<SEED_SIZE, 16> + 'a,
+        RA: ReportAggregationUpdate + Clone + 'a;
+
+    /// Returns a flag indicating whether aggregation jobs are being created or updated.
+    fn creating_aggregation_jobs(&self) -> bool;
```

Comment:
nit: I think this would be slightly easier to follow if we returned an enum rather than a bool -- something like `AggregationJobWriteType{ Create, Update }`. Otherwise, the reader has to remember that `true` is creation, `false` is update.